### ISSUE TITLE
Question: Read Contents security concern

### ISSUE THREAD

**Issue by amazzotti** [atlassian/github-for-jira#565]
Hello there, first of all thank you for this fantastic app!

I have a security concern however enabling this to my organization. I would like to enable it for all repositories, so that all users will benefit from it, however the App requires "Read-only access to Contents". 

As far I understand the integration only needs repository metadata, however this permission will also grant read access to the [GET /repos/{owner}/{repo}/contents/{path}](https://docs.github.com/en/rest/reference/repos#get-repository-content) endpoint. 

If this is true, an attacker with the App installation token will be able to read all source code of all repositories within the entire organization. Do I understand this correctly?

As far I can read the documentation, this seems to be a limitation within GitHub permissions not being granular enough. There seems to be no way to only grant access to repository metadata, without exposing the entire source code. 

Do you have any suggestion on how to avoid or mitigate this risk?  
Thank you a lot!

**Comment by mboudreau** [comment]
Hi!

Thanks for this very good security question!  I'll be very happy to try to answer any concerns you might have.

First, let's tackle the permission itself (Read-Only Contents) and why we have it. The [Github docs](https://docs.github.com/en/rest/reference/permissions-required-for-github-apps#permission-on-contents) describe is as having access to:
> Repository contents, commits, branches, downloads, releases, and merges.
This is crucial for our app as the whole functionality revolves around commits, branches, releases, merges, etc.  We cannot change it as the app would lose it's purpose without it.  However, it would be nice to have more granular permissions (like you mentioned) so that we don't have specifically code access, just the commits, branches, etc, but you would need to bring this up with Github. I personally doubt they actually can change this without affecting a great number of systems, but it's always worth a try.

As for Attackers getting in and getting access to your code, several _major_ things would have to go wrong for this to happen as the attacker would need several pieces of information and access to multiple systems to gain access.  We're still in  the middle of migrating the ownership from Github to Atlassian, but I can tell you that from the Atlassian side that we are using the best security practices available like least privilege access roles, encryption at rest for the database/redis and our keys, no in-memory caching within the app (transient data only), no tokens stored across requests, secure cookies, CSRF, etc.  

We take security very seriously at Atlassian and we share the same concerns that you have.  We are also using the app ourselves just to show how much trust we have in it.  It's also one of the core reason we believe in having the repository Open Source to show that we're not hiding anything.  [Feel free to check out the micros branch](https://github.com/atlassian/github-for-jira/tree/micros) that we've been working on which will become the new main once the migration is done.

Now to get a bit more philosophical, Security will always be at odds with Convenience.  The most secure system is the one that isn't even turned on, but that's not very useful.  We need to strike the right balance between the two, but that means introducing some risk in the equation and thinking about worse case scenarios. What would be the worst case for you?  Someone having full access to the code?  Is that really a big issue?  

From my experience, if you're not comfortable with having your code looked at by a potential intruder, it's probably because you're using bad security practices already, like having secrets in the codebase and other sensitive information.  If that's the case, I'd recommend you don't use this app and that you should probably fix that issue because there's other ways the code can get out, like how every developer on the repository has a copy of the whole codebase on their computers. Security by obfuscation isn't very secure. If the issue here is more around Intellectual Property, then there isn't much I can do for you (see developers having copies of the code everywhere) and the real value isn't the software because software without being deployed or utilized doesn't have any value - the real value comes from how the software is used in the world which is extremely hard to replicate by a malicious actor.  Most just try to get sensitive information like databases as that's where the value lies.  

Anywho, sorry for the short novella, I guess the question comes down to: is the software you have need to be so secure that it's okay to lose many developer-hours managing Jira tickets with Github Commits/PRs/Releases?

**Comment by amazzotti** [comment]
Thank you for your extensive reply!

Just to clarify, we put a lot of trust in the app, since as you mentioned it's open source and we can easily audit the code, and again thank you for your amazing work.

To elaborate more on our context, we have a big GitHub organization with hundreds of repositories. At the same time we have several Jira Cloud instances that want to integrate with this organization. They are ran by different departments, and they only need integration on some subset of the repositories (the ones relevant to them).
So as organization admin, my goal is to implement least privilege access on these repositories.

One mitigation on our side would be to create different GitHub organizations, so that they can map 1:1 with the Jira Cloud instances. However this adds a bigger administrative burden to us and also this mapping is not always 1:1 or consistent with other app integrations, so there will be always some overlapping.
We've been on a similar case with the Azure DevOps Pipelines app, where any developer could craft a malicious pipeline and clone (and write to) any repository in the organization, even the ones they personally don't have access to.  

For this reason we had to avoid the app and enable integration through Personal Access Tokens. In our context each product team has ownership of one or more "Service Accounts" (just a normal GitHub user), so that they can integrate granularly and autonomously using PATs.
We already asked GitHub to be able to enable Apps per repository level, and not the entire organization, so that we can use different installation tokens per repository (or per repository subset ideally). This is however is quite a big feature request and it does have some impact over the conveniency of GitHub Apps.

You are completely right on the risks coming mostly from secret leaking in the source code, which is also something that we can and should absolutely avoid.
As you mentioned however the issue in this case is about intellectual property, and while I share your same opinion, not all the project leads in the organization think the same and they might refuse using our GitHub organization over security concerns.
We also trust Atlassian to store secrets (in this case the application credentials) in a secure way. 

My concern would be way bigger if we would need to integrate on Jira self-hosted instances, because then as GitHub organization admin I would have to trust several different parties and it would be possible for each of them to steal the credentials from Jira. Anyway, this is not the case here, so I agree with you many things have to go wrong for this scenario to happen, and no disgruntled Jira organization admin will have access to the app credentials anyway since they are safe "on the cloud".

This is not a hard request, more of an idea, but would it be possible on your side to allow integration through Personal Access Tokens? This would allow the organization admin to create PATs with the ```repo``` scope and implement granularity by having several "Service Accounts" that map 1:1 with the Jira Cloud instances. The repository administrator then can simply add the "Service Account" relevant to them with Read access to the repository. It's a bit of a cumbersome process, but this is how we are enabling most integrations nowadays in our organization. The PAT would still have read access to the source code, and the PAT would have more chances to leak if not handled properly, so maybe this is not the best idea ever, but it would at least allow to scope it to only a subset of repositories.

For an actual solution instead, this can indeed be translated to a GitHub feature request.
The permission system can be more granular and the contents (and tree?) endpoint access can be split into a different permission scope.
I could link this issue to it, since I believe it makes an excellent use case.

I'm closing this issue since as we analyzed this is a limitation of the GitHub permission system and not the application.  It is also not possible to exploit the app to read repository content, since the app does never read content, and it's unlikely for an attacker to access the application credentials, since they are safely stored on the cloud and not accessible.   

Thank you a lot for your time and helping me elaborating this.

**Comment by mboudreau** [comment]
Let's keep this open for now, I think it's important to address the concerns you're bringing up.

So, while I was contemplating this, an idea hit me - seems like you need Github Enterprise Server.  The Github team are currently working on a way to integration our app with it in a way that's hyper secure which I'm fairly sure would involve some kind of GPG public/private key specific to the organization so that even if Atlassian got breached, the data would be completely useless without the private key, which only your organization would hold.

Of course, this will cost more but as is most things if you want it secure ;)

If this sounds like an avenue you'd like to explore, I'd be happy to talk to the Github team about it moving forward to see what we can both do to make this work for you.

Cheers.

**Comment by amazzotti** [comment]
We have a preference on GitHub over GitHub Enterprise Server. Your suggestion is good to re-evaluate this choice, but I believe for the foreseeable future we will stick to GitHub.com  

I submitted a feedback to GitHub asking if it would be possible to decouple the ```contents``` access on a different scope level. I believe this would benefit the many applications which only need access to metadata, rather than source code, for example to monitor commits, generate metrics, reports, etc.  

I also found [this](https://github.com/github/roadmap/issues/184) issue in GitHub's roadmap. It's a bit vague, but it does mention _"Fine-grained permissions across specific features"_, so could it cover this use case?  

It also mentions Personal Access Tokens (PAT) only, but I assume the same fine-grained system will be applied to GitHub Apps and OAuth Apps.

**Comment by rachellerathbone** [comment]
Hey @amazzotti. Did you hear anything back from GitHub?

**Comment by rachellerathbone** [comment]
Also, just wanted to let you know that we've changed our permissions so we now no longer require any access to actions and only read permissions for deployments. I've also updated the permissions section of our [README](https://github.com/atlassian/github-for-jira/blob/main/README.md) and added an [FAQs](https://github.com/atlassian/github-for-jira/blob/main/docs/FAQs.md) doc that the team plans to add to when we see questions being commonly asked by our users.

**Comment by amazzotti** [comment]
Thank you @rachellerathbone for the notice. 
I don't have updates from GitHub.

Great to see the FAQ, I think that really helps!
I would slightly change the last sentence of the first reply though, to:
"Our app simply sends metadata through to Jira, no repository content is read during this process."

**Comment by rachellerathbone** [comment]
As this is resolved, going to go ahead and close this one out 🎉
